Fix the incorrect enabled state after removing a disabled program from the programs table

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #762 

Changes to the programs table on the Dashboard page

- Fix the incorrect enabled state after removing a disabled program
- Bonus: cancel the focus from the "Remove" links after removing a program

### Screenshots:

#### The enabled states are correct after the remove

![Kapture 2021-06-09 at 16 03 32](https://user-images.githubusercontent.com/17420811/121318927-4a802700-c93e-11eb-8bc4-04380e0fc6b1.gif)

#### Cancel the focus

![image](https://user-images.githubusercontent.com/17420811/121319900-3be63f80-c93f-11eb-9d06-405d15e242a4.png)

In the first demo GIF, the focus is locating to the next "Remove" link after the remove. This demo GIF shows that the https://github.com/woocommerce/google-listings-and-ads/commit/82975fe46348c6fce0b8ab1f890b3ee4ccc2301f was to fix this minor issue.

![Kapture 2021-06-09 at 16 03 40](https://user-images.githubusercontent.com/17420811/121319756-19542680-c93f-11eb-94b2-e540a59314ae.gif)

### Detailed test instructions:

1. Create at least two paid programs
2. Disabled the program at the top of the programs table and then remove it
3. The enabled state of the program, which is under that removed program, should be the correct state

### Changelog entry

- The incorrect enabled state after removing a disabled program from the programs table
- Cancel the focus from the "Remove" links after removing a program from the programs table

eason9487 committed:
Fix incorrect enabled state after removing a disabled program from programs table

eason9487 committed:
Cancel the focus from Remove links after removing a program from programs table

ecgan commented on js/src/dashboard/all-programs-table-card/index.js:
```diff
@@ -113,15 +113,18 @@ const AllProgramsTableCard = ( props ) => {
 					{
 						display: (
 							<div className="program-actions">
 								<EditProgramButton programId={ el.id } />
 								{ el.id !== FREE_LISTINGS_PROGRAM_ID && (
-									<RemoveProgramButton programId={ el.id } />
+									<RemoveProgramButton
+										key={ el.id }
+										programId={ el.id }
+									/>
 								) }
 							</div>
```
💅  Instead of putting `key={ el.id }` at the `RemoveProgramButton`, I would put it at the top-most `div` wrapping everything inside.

eason9487 replied:
Moved by https://github.com/woocommerce/google-listings-and-ads/pull/771/commits/77848d9714f4ecff3be1150be85ed144a6c5a5ea and added comments for this fix.

ecgan reviewed (approved):
LGTM, tested locally, seems working good. 👍  Just left one minor nitpick comment.

ecgan commented:
Just FYI for reference, I was wondering what causes the issue and why specifying the `key` would solve the issue. Turns out, I think it was because of `@woocommerce/components`'s `Table` component rendering rows using array index as keys, which could give us this kind of problem when adding/removing rows.

https://github.com/woocommerce/woocommerce-admin/blob/66cdf344ea615e451428bb8c419f974cf413a44a/packages/components/src/table/table.js#L288-L289

eason9487 commented:
> Turns out, I think it was because of `@woocommerce/components`'s `Table` component rendering rows using array index as keys, which could give us this kind of problem when adding/removing rows.

Indeed. In most cases, the array indices aren't the proper keys when rendering component array. And that's one of the examples of misuse.

eason9487 committed:
Move key prop to the upper level and add relevant comments

Address https://github.com/woocommerce/google-listings-and-ads/pull/771#discussion_r648427669
